src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,251 @@
+        Map<String, Long> umiFrequencies = records.stream()
+                        .collect(Collectors.groupingBy(p -> p.getStringAttribute(umiTag),
+                                counting()));
+
+        UmiGraph umiGraph = new UmiGraph();
+
+        // Add each of the uniquely observed UMI sequences to the graph
+        int firstGroup = 0;
+        for (Map.Entry<String, Long> umiFrequency : umiFrequencies.entrySet()) {
+            umiGraph.addNode(new UmiSequence(umiFrequency.getKey(), umiFrequency.getValue()));
+        }
+        umiGraph.processNodes();
+
+        Map<Integer, List<UmiSequence>> umisByDuplicateSet = umiGraph.getUmisByDuplicateSet();
+
+        // Construct DuplicateSetList
+        List<DuplicateSet> duplicateSetList = new ArrayList<>();
+        for(int i = 0;i < umiGraph.duplicateSetCounter;i++) {
+            DuplicateSet e = new DuplicateSet();
+            duplicateSetList.add(e);
+        }
+
+        for(SAMRecord rec : records) {
+            String umi = records.get(records.indexOf(rec)).getStringAttribute(umiTag);
+
+            // Figure out which group this belongs to
+            int recordGroup = umiGraph.node.get(umi).duplicateSet;
+            duplicateSetList.get(recordGroup).add(records.get(records.indexOf(rec)));
+        }
+
+        // Optionally add the inferred Umi
+        if(addInferredUmi) {
+            for(DuplicateSet ds : duplicateSetList) {
+                // For each duplicate set identify the most common UMI
+                List<SAMRecord> recordsFromDuplicateSet = ds.getRecords();
+
+                // Assign the inferred UMI to all reads in the current group
+                for(SAMRecord record : recordsFromDuplicateSet) {
+                    record.setAttribute(inferredUmiTag, umiGraph.getInferredUmiFromOriginalSequence(record.getStringAttribute(umiTag)));
+                }
+            }
+        }
+
+        nextSetsIterator = duplicateSetList.iterator();
+    }
+
+    private int getEditDistance(final String s1, final String s2) {
+        if(s1 == null || s2 == null) {
+            throw new PicardException("Attempt to compare two incomparable UMIs.  At least one of the UMIs was null.");
+        }
+        if(s1.length() != s2.length()) {
+            throw new PicardException("Barcode " + s1 + " and " + s2 + " do not have matching lengths.");
+        }
+        int count = 0;
+        for(int i = 0;i < s1.length();i++) {
+            if(s1.charAt(i) != s2.charAt(i)) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    private class UmiSequence {
+        private final String sequence;
+        private final Long occupancy;
+        private String inferredSequence;
+        private Integer duplicateSet;
+
+        UmiSequence(final String sequence, final Long occupancy) {
+            this.sequence = sequence;
+            this.occupancy = occupancy;
+        }
+    }
+
+    private class UmiGraph {
```

Comment:
extract this class please. you can make it package local if you don't want others to use it, but right now you are accessing private members....which is not needed and good to avoid.